File: implement-shell-tools/wc/wc.js
const fs = require("fs");

const args = process.argv.slice(2);

let countLines = false;
let countWords = false;
let countBytes = false;
let files = [];

for (const arg of args) {
    if (arg === "-l") {
        countLines = true;
    } else if (arg === "-w") {
        countWords = true;
    } else if (arg === "-c") {
        countBytes = true;
    } else {
        files.push(arg);
    }
}

// If no flags are given, wc shows all three
if (!countLines && !countWords && !countBytes) {
    countLines = true;
    countWords = true;
    countBytes = true;
}

function countFile(filename) {
     try {
        const content = fs.readFileSync(filename, "utf8");

        const lines = content.split("\n").length - 1;
        const words = content.trim() === "" ? 0 : content.trim().split(/\s+/).length;
        const bytes = Buffer.byteLength(content);

        let output = [];

        if (countLines) {
            output.push(lines);
        }

        if (countWords) {
            output.push(words);
        }

        if (countBytes) {
            output.push(bytes);
        }

        output.push(filename);

        console.log(output.join(" "));

        return {
            lines,
            words,
            bytes
        };
    } catch (err) {
        console.error(`wc: ${filename}: ${err.message}`);
        return null;
    }
}

let totalLines = 0;
let totalWords = 0;
let totalBytes = 0;

for (const file of files) {
    const counts = countFile(file);

    if (counts) {
        totalLines += counts.lines;
        totalWords += counts.words;
        totalBytes += counts.bytes;
    }
}

if (files.length > 1) {
    let output = [];

    if (countLines) {
        output.push(totalLines);
    }

    if (countWords) {
        output.push(totalWords);
    }

    if (countBytes) {
        output.push(totalBytes);
    }

    output.push("total");

    console.log(output.join(" "));
}

File: implement-shell-tools/wc/wc.py
import sys

args = sys.argv[1:]

count_lines = False
count_words = False
count_bytes = False
files = []

for arg in args:
    if arg == "-l":
        count_lines = True
    elif arg == "-w":
        count_words = True
    elif arg == "-c":
        count_bytes = True
    else:
        files.append(arg)

# If no flags are given, wc shows all three
if not count_lines and not count_words and not count_bytes:
    count_lines = True
    count_words = True
    count_bytes = True

def print_output(lines, words, bytes_count, label):
    output = []

    if count_lines:
        output.append(f"{lines:8}")

    if count_words:
        output.append(f"{words:8}")

    if count_bytes:
        output.append(f"{bytes_count:8}")

    output.append(f" {label}")

    print("".join(output))


def count_file(filename):
    try:
        with open(filename, "r") as file:
            content = file.read()

        lines = content.count("\n")

        words = 0 if content.strip() == "" else len(content.split())

        # UTF-8 bytes, same idea as Buffer.byteLength()
        bytes_count = len(content.encode("utf-8"))

        print_output(lines, words, bytes_count, filename)

        return {
            "lines": lines,
            "words": words,
            "bytes": bytes_count
        }

    except Exception as err:
        print(f"wc: {filename}: {err}")
        return None


total_lines = 0
total_words = 0
total_bytes = 0



for file in files:
    counts = count_file(file)

    if counts:
        total_lines += counts["lines"]
        total_words += counts["words"]
        total_bytes += counts["bytes"]


if len(files) > 1:
    print_output(total_lines, total_words, total_bytes, "total")